File: src/models/room.py
from .item import Item


class Room:

  def __init__(self, name, description, inventory=None):
    self.name = name
    self.description = description
    self.inv = inventory if inventory else []

  def remove_from_inv(self, name):
    """
    If possible, removes item from room.
    Casing does not matter.

    If the item is found, removes the item 
    from the inventory and returns the item.

    Else, returns None.
    """
    for item in self.inv:
      if item.name.lower() == name:
        self.inv.remove(item)
        return item
    return None

  def add_to_inv(self, item):
    """Adds item to room."""
    self.inv.append(item)


class DebugRoom(Room):
  def __init__(self):
    name = "|——————————————————————————————————————————————————————————————————————————————|"
    desc = "This adventure game requires a terminal 80 characters wide and 15 lines tall."
    inventory = [Item("Please resize"), Item(
        "So you can see"), Item("This text")]
    super().__init__(name, desc, inventory)

File: JS/src/models/Room.js
const { Item } = require('./Item');
const { Guardian } = require('./Guardians');

class Room {
    /**
     *
     * @param {string} name
     * @param {string} description
     * @param {Item[]} inventory
     */
    constructor(name, description, inventory = []) {
        this.name = name;
        this.description = description;
        this.inventory = inventory;
    }

    /**
     * If possible, removes item from room.
     * Casing does not matter.
     *
     * @param {string} name
     *
     * @returns {null | Item} The item removed from the room inventory, or null.
     */
    remove_from_inventory(name) {
        let item = null;
        const itemIndex = this.inventory.findIndex((item) => item.name.toLowerCase() === name);
        if (itemIndex > -1) {
            item = this.inventory[itemIndex];
            this.inventory.splice(itemIndex, 1);
        }
        return item;
    }

    /**
     *
     * @param {Item} item
     */
    add_to_inventory(item) {
        this.inventory.push(item);
    }
}
class Shrine extends Room {
    /**
     *
     * @param {string} name
     * @param {string} description
     * @param {Item[]} inventory
     * @param {Guardian | null} guardian
     */
    constructor(name, description, inventory = [], guardian = null) {
        super(name, description, inventory);
        this.__guardian = guardian;
        if (this.__guardian != null) {
            this.__guardian.shrine = this;
        }
    }

    get guardian() {
        return this.__guardian;
    }

    set guardian(guardian) {
        if (guardian) {
            guardian.shrine = this;
        }
        this.__guardian = guardian;
    }
}

const debugRoom = new Room(
    "You're an explorer, you see!",
    'Search through the Mathematical Caves, find the Artifact, and make your escape!',
    [new Item('Please resize'), new Item('So you can see'), new Item('All of'), new Item('This text')]
);

module.exports = {
    Room,
    Shrine,
    debugRoom,
};
